File: ten-basic-algorithms/chunkArray/chunk-array.go
package main

import "fmt"

func chunkArray[T any](arr []T, size int) [][]T {
	// Empty slice to hold the chunks
	chunks := make([][]T, 0)
	// Loop through the array in steps of size
	for i := 0; i < len(arr); i += size {
		// Calculate the end index for the chunk
		end := i + size

		// If the end index exceeds the length of the array, set it to the length of the array
		if end > len(arr) {
			end = len(arr)
		}

		// Append the chunk to the chunks slice
		chunks = append(chunks, arr[i:end])
	}

	return chunks
}

func main() {
	arr := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
	size := 3
	chunks := chunkArray(arr, size)
	fmt.Println(chunks) // Output: [[1 2 3] [4 5 6] [7 8 9]]

	// Test with a string array
	strArr := []string{"a", "b", "c", "d", "e", "f", "g", "h"}
	strSize := 3
	strChunks := chunkArray(strArr, strSize)
	fmt.Println(strChunks) // Output: [[a b c] [d e f] [g h]]
}

File: basic-algorithms/chunkArray/chunk-array.ts
function chunkArray<T>(arr: T[], size: number): T[][] {
  // Create an empty array to hold the chunks
  const result: T[][] = [];
  // Loop through the array and slice it into chunks of the specified size
  for (let i = 0; i < arr.length; i += size) {
    // Use the slice method to create a chunk and push it to the result array
    // The slice method takes two arguments: the start index and the end index (exclusive)
    result.push(arr.slice(i, i + size));
  }
  // Return the array of chunks
  return result;
}